File: semester 3/sortingAndPencarianBiner.cpp
#include <iostream>

using namespace std;

void tampilArray(int n, int arr[]) {
    int i;
    for (i = 0; i < n; i++) {
        cout << arr[i] << " ";
    }
    cout << endl;
}

void bubbleSort(int n, int arr[]) {
    int i, j, temp;
    bool tukar;

    for (i = 0; i < n - 1; i++) {
        tukar = false;
        for (j = 0; j < n - i - 1; j++) {
            if (arr[j] > arr[j + 1]) {
                temp = arr[j];
                arr[j] = arr[j + 1];
                arr[j + 1] = temp;
                tukar = true;
            }
        }
        if (!tukar) {
            break;
        }
        tampilArray(n, arr); // Menampilkan array setelah setiap perulangan
    }
}

void insertionSort(int n, int arr[]) {
    int i, j, key;
    for (i = 1; i < n; i++) {
        key = arr[i];
        j = i - 1;

        while (j >= 0 && arr[j] > key) {
            arr[j + 1] = arr[j];
            j = j - 1;
        }
        arr[j + 1] = key;
        tampilArray(n, arr); // Menampilkan array setelah setiap perulangan
    }
}

void selectionSort(int n, int arr[]) {
    int i, j, minIndex, temp;
    for (i = 0; i < n - 1; i++) {
        minIndex = i;
        for (j = i + 1; j < n; j++) {
            if (arr[j] < arr[minIndex]) {
                minIndex = j;
            }
        }
        temp = arr[minIndex];
        arr[minIndex] = arr[i];
        arr[i] = temp;
        tampilArray(n, arr); // Menampilkan array setelah setiap perulangan
    }
}

int binarySearch(int arr[], int low, int high, int cari) {
    while (low <= high) {
        int tengah = low + (high - low) / 2;

        if (arr[tengah] == cari) {
            return tengah;
        }
        else if (arr[tengah] < cari) {
            low = tengah + 1;
        }
        else {
            high = tengah - 1;
        }
    }

    return -1;
}

int main() {
    int n, i;
    cout << "Dikoding oleh\nNama: Indra Dwi Aryadi (xxxxxxxxxxxx)\nKelas: 03TPLP003"<<endl;
    cout << "Masukkan jumlah elemen: ";
    cin >> n;

    int tabInt[n];
    cout << "Masukkan elemen-elemen array: ";
    for (i = 0; i < n; i++) {
        cin >> tabInt[i];
    }

    cout << "Array sebelum diurutkan: ";
    tampilArray(n, tabInt);

    int bubbleSortArr[n];
    for (i = 0; i < n; i++) {
        bubbleSortArr[i] = tabInt[i];
    }
    cout << "Bubble Sort:\n";
    bubbleSort(n, bubbleSortArr);

    int insertionSortArr[n];
    for (i = 0; i < n; i++) {
        insertionSortArr[i] = tabInt[i];
    }
    cout << "Insertion Sort:\n";
    insertionSort(n, insertionSortArr);

    int selectionSortArr[n];
    for (i = 0; i < n; i++) {
        selectionSortArr[i] = tabInt[i];
    }
    cout << "Selection Sort:\n";
    selectionSort(n, selectionSortArr);

    cout << "\n\n";
    cout << "Dikoding oleh\nNama: Indra Dwi Aryadi (xxxxxxxxxxx)\nKelas: 03TPLP003"<<endl;
    cout << "Pencarian Biner:\n";
    int cari;
    cout << "Masukkan angka yang ingin dicari: ";
    cin >> cari;

    // Pencarian pada array yang telah diurutkan menggunakan binary search
    int posisi = binarySearch(bubbleSortArr, 0, n - 1, cari);
    if (posisi != -1) {
        cout<< "Data telah diurutkan dan pencarian biner: "<<endl;
        cout << "Angka " << cari << " ditemukan pada indeks ke-" << posisi << endl;
    } else {
        cout << "Angka " << cari << " tidak ditemukan" << endl;
    }

    return 0;
}

File: semester 3/sortingAndPencarianbiner.py
def tampil_array(arr):
    for elemen in arr:
        print(elemen, end=" ")
    print()

def bubble_sort(arr):
    n = len(arr)
    for i in range(n - 1):
        tukar = False
        for j in range(n - i - 1):
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
                tukar = True
        if not tukar:
            break
        tampil_array(arr)

def insertion_sort(arr):
    n = len(arr)
    for i in range(1, n):
        key = arr[i]
        j = i - 1
        while j >= 0 and arr[j] > key:
            arr[j + 1] = arr[j]
            j = j - 1
        arr[j + 1] = key
        tampil_array(arr)

def selection_sort(arr):
    n = len(arr)
    for i in range(n - 1):
        min_index = i
        for j in range(i + 1, n):
            if arr[j] < arr[min_index]:
                min_index = j
        arr[min_index], arr[i] = arr[i], arr[min_index]
        tampil_array(arr)

def binary_search(arr, low, high, cari):
    while low <= high:
        tengah = low + (high - low) // 2
        if arr[tengah] == cari:
            return tengah
        elif arr[tengah] < cari:
            low = tengah + 1
        else:
            high = tengah - 1
    return -1

print("Dikoding oleh\nNama: Indra Dwi Aryadi (xxxxxxxxxx)\nKelas: 03TPLP003")
n = int(input("Masukkan jumlah elemen: "))
tab_int = []
print("Masukkan elemen-elemen array: ")
for _ in range(n):
    elemen = int(input())
    tab_int.append(elemen)

print("Array sebelum diurutkan: ", end="")
tampil_array(tab_int)

bubble_sort_arr = tab_int.copy()
print("Bubble Sort:")
bubble_sort(bubble_sort_arr)

insertion_sort_arr = tab_int.copy()
print("Insertion Sort:")
insertion_sort(insertion_sort_arr)

selection_sort_arr = tab_int.copy()
print("Selection Sort:")
selection_sort(selection_sort_arr)

print("\n\n")
print("Dikoding oleh\nNama: Indra Dwi Aryadi (xxxxxxxx)\nKelas: 03TPLP003")
print("Pencarian Biner:")
cari = int(input("Masukkan angka yang ingin dicari: "))

# Pencarian pada array yang telah diurutkan menggunakan binary search
posisi = binary_search(bubble_sort_arr, 0, n - 1, cari)
if posisi != -1:
    print("Data telah diurutkan dan pencarian biner:")
    print(f"Angka {cari} ditemukan pada indeks ke-{posisi}")
else:
    print(f"Angka {cari} tidak ditemukan")
